### ISSUE TITLE
How update nested object with relationships

### ISSUE THREAD

**Issue by Jenan** [dotnet/efcore#9819]
Hello,

I have tried to update the following tables structure:

Tables:
```C#
public class Resource
  {
    public int Id { get; set; }

    public string Name { get; set; }

    public List<Scope> Scopes { get; set; }
  }

public class Scope
  {
    public int Id { get; set; }

    public string Name { get; set; }

    public List<Claim> Claims { get; set; }

    public Resource Resource { get; set; }
  }


public class Claim
  {
    public int Id { get; set; }

    public string Type { get; set; }

    public Scope Scope { get; set; }
  }
```
I have a `Resource`:
```
Id: 1
Name: "TEST"
Scopes: [
 Id: 1,
 Name: "TEST-SCOPE",
 Claims: [
 Id: 1,
 Type: "TEST-CLAIM"
]
]
```

I would like to update Scope: ID = 1

I have tried this:
```C#
public async Task<int> UpdateScopeAsync(Scope scope)
{
            _dbContext.Scopes.Update(scope);
            
            var result = await _dbContext.SaveChangesAsync();
            
            return result;
}
```
Where scope param cotains following structure:
scope: 
```
Id = 1, 
Name = "TEST-SCOPE-2",
Claims: [
 Id: 1,
 Type: "TEST-CLAIM-2"
], 
Resource = {
 Id = 1
}
```

I get this error:

> The instance of entity type 'Scope' cannot be tracked because another instance of this type with the same key is already being tracked. When adding new entities, for most key types a unique temporary key value will be created if no key is set (i.e. if the key property is assigned the default value for its type). If you are explicitly setting key values for new entities, ensure they do not collide with existing entities or temporary values generated for other new entities. When attaching existing entities, ensure that only one entity instance with a given key value is attached to the context.

How can I fix this please?

Thank you.

**Comment by ajcvickers** [comment]
@Jenan There isn't enough information here for us to be able to reproduce what you are seeing. Have you tried debugging to see whether the duplicate instance is coming from? If you would like us to do more investigation, then please post a complete listing or project that reproduces what you are seeing.

**Comment by Jenan** [comment]
@ajcvickers Hello, I have created this sample:

https://github.com/Jenan/EfCoreTesting

Please, take a look. :) How can I fix that?

**Comment by ajcvickers** [comment]
@Jenan Fundamentally, the issue is what the exception message says. The code creates multiple instances of Client and ClientScope with the same key values. Instead, the code needs to determine whether or not an instance already exists, and then use it if it does. A common way to do this is to query the database for entities and then make updates to the entities returned before calling SaveChanges.

Looking specifically at your test code, if I stop creating new instances for the same entity:
```C#
//Try add Client
var newClient = new Client() { ClientId = "Test", ClientName = "Test" }; //Primary key is Id, not ClientId
dbContext.Clients.Add(newClient);
dbContext.SaveChanges();

//Try add ClientScope for this client Id = 1
var newClientScope = new ClientScope() { Client = newClient, Scope = "Test Scope" };
dbContext.ClientScopes.Add(newClientScope);

dbContext.SaveChanges();

//Try update ClientScope name
newClientScope.Scope = "Test Scope 2";
dbContext.SaveChanges();
```
then it works as expected.

**Comment by Jenan** [comment]
Thank you, I have a question about updating ClientScope - if I want to update ClientScope and I know only client id - can I update the ClientScope with new Client() { Id = 1 } - or is it necesarry get whole client object before updating ClientScope? That is point which I don't understand correctly - because from form I get only ClientId and ClientScope information. And I would like to write some base repository with update method but it seems that is not possible write common update method if you have a related data - what do you recommend? Thank you.

**Comment by ajcvickers** [comment]
@Jenan There are various ways to do it, but usually the easiest is to do a query like you show in your second code block above.

**Comment by Jenan** [comment]
@ajcvickers Thank you - what do you prefer to use for base repository for updating method? There it will be hard specify this relation, what do you think about it?

**Comment by ajcvickers** [comment]
@Jenan I don't fully understand your question, but in general I would only add a repository on top of the context if there is need for an additional layer of abstraction, and in those cases usually using a generic repository doesn't make the abstraction clean enough. But these things are very much dependent on your application and your preferences as a developer.

**Comment by Jenan** [comment]
OK, I see. Thank you for your suggestion. I've posted this type of question on SO and I got a great advice about model structure above. In EF model especially in ClientScope missing the foreign key like this:

```cs
public class ClientScope
{
    public int Id { get; set; }

    public string Scope { get; set; }

    public Client Client { get; set; }
   //This foreign key -> 
   public int ClientId { get; set; }
}
```

If I have a ClientId foreign key in ClientScope - I can't load whole Client object and it is possible update ClientScope like this: 
new ClientScope(){ Scope = "Test", Id = 1, ClientId = 1 } and works perfectly. :)
